src/uct/tcp/tcp_cm.c

```diff
@@ -8,86 +8,329 @@
+
+    switch(ep->conn_state) {
+    case UCT_TCP_EP_CONN_STATE_CONNECTING:
+    case UCT_TCP_EP_CONN_STATE_WAITING_ACK:
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) {
+            uct_tcp_iface_outstanding_inc(iface);
+        } else {
+            ucs_assert(old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CONNECTED:
+        ucs_assert((old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
+                   (old_conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING));
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CLOSED:
+        ucs_assert(old_conn_state != UCT_TCP_EP_CONN_STATE_CLOSED);
+        if ((old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
+            (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK)) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    default:
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING);
+        break;
+    }
+
+    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
+              ep, uct_tcp_ep_cm_state[old_conn_state].name,
+              uct_tcp_ep_cm_state[ep->conn_state].name,
+              ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
+                               str_local_addr, UCS_SOCKADDR_STRING_LEN),
+              ucs_sockaddr_str(ep->peer_addr->addr, str_remote_addr,
+                               UCS_SOCKADDR_STRING_LEN));
+}
+
+static void uct_tcp_cm_trace_conn_pkt(const uct_tcp_ep_t *ep, const char *msg,
+                                      const struct sockaddr *peer_addr)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+
+    ucs_debug("tcp_ep %p: %s %s", ep, msg,
+              ucs_sockaddr_str(peer_addr, str_addr,
+                               UCS_SOCKADDR_STRING_LEN));
+}
+
+static ucs_status_t
+uct_tcp_cm_conn_pkt_check_event(const uct_tcp_ep_t *ep,
+                                uct_tcp_cm_conn_event_t expected_event,
+                                uct_tcp_cm_conn_event_t actual_event)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_assertv_always(expected_event == actual_event,
+                       "tcp_ep %p: received wrong CM event (expected: %u, "
+                       "actual: %u) from the peer with iface listener "
+                       "address: %s", ep, expected_event, actual_event,
+                       ucs_sockaddr_str(ep->peer_addr->addr, str_addr,
+                                        UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static void uct_tcp_cm_log_somaxconn_exceed(ucs_log_level_t log_level)
+{
+    ucs_log(log_level, "try to increase the accept queue length: "
+            "\"sysctl -w net.core.somaxconn=<new value>\" where "
+            "<new_value> greater than %d", ucs_socket_max_conn());
```

Comment:
what if ucs_socket_max_conn() could not read the file from /proc? then the error message seems irrelevant, right?

Reply:
but then what is the meaning to set "sysctl -w net.core.somaxconn" to a greater value?
if /proc/.. does not exist, the there is nothing to do; we should write that the hard limit is "%d",SOMAXCONN and that's it

Reply:
Added function that always reads somaxconn value from the system path, and use it here to understand whether we can recommend increasing the `net.core.somaxconn` value or not.
@yosefe  wdyt?